shared/tinyusb/mp_usbd.h

```diff
@@ -27,25 +27,117 @@
 
+#include "py/mpconfig.h"
+
+#if MICROPY_HW_ENABLE_USBDEV
+
 #include "py/obj.h"
-#include "tusb.h"
+#include "py/objarray.h"
+#include "py/runtime.h"
 
-static inline void mp_usbd_init(void) {
-    // Currently this is a thin wrapper around tusb_init(), however
-    // runtime USB support will require this to be extended.
-    tusb_init();
-}
+#ifndef NO_QSTR
+#include "tusb.h"
+#include "device/dcd.h"
+#endif
 
-// Call this to explicitly run the TinyUSB device task.
+// Run the TinyUSB device task
 void mp_usbd_task(void);
 
+// Schedule a call to mp_usbd_task(), even if no USB interrupt has occurred
+void mp_usbd_schedule_task(void);
+
 // Function to be implemented in port code.
 // Can write a string up to MICROPY_HW_USB_DESC_STR_MAX characters long, plus terminating byte.
 extern void mp_usbd_port_get_serial_number(char *buf);
 
-// Most ports need to write a hexadecimal serial number from a byte array, this
+// Most ports need to write a hexadecimal serial number from a byte array. This
 // is a helper function for this. out_str must be long enough to hold a string of total
 // length (2 * bytes_len + 1) (including NUL terminator).
 void mp_usbd_hex_str(char *out_str, const uint8_t *bytes, size_t bytes_len);
 
+// Length of built-in configuration descriptor
+#define MP_USBD_BUILTIN_DESC_CFG_LEN (TUD_CONFIG_DESC_LEN +                     \
+    (CFG_TUD_CDC ? (TUD_CDC_DESC_LEN) : 0) +  \
+    (CFG_TUD_MSC ? (TUD_MSC_DESC_LEN) : 0)    \
+    )
+
+// Built-in USB device and configuration descriptor values
+extern const tusb_desc_device_t mp_usbd_builtin_desc_dev;
+extern const uint8_t mp_usbd_builtin_desc_cfg[MP_USBD_BUILTIN_DESC_CFG_LEN];
+
+void mp_usbd_task_callback(mp_sched_node_t *node);
+
+#if MICROPY_HW_ENABLE_USB_RUNTIME_DEVICE
+void mp_usbd_deinit(void);
+void mp_usbd_init(void);
+
+const char *mp_usbd_runtime_string_cb(uint8_t index);
+
+// Provided for future expansion, currently only the default built-in driver
+// configuration (or none) can be set.
+typedef enum {
+    MP_USBD_BUILTIN_NONE,
+    #if CFG_TUD_CDC && !CFG_TUD_MSC
+    MP_USBD_BUILTIN_CDC,
+    #endif
+    #if CFG_TUD_MSC && !CFG_TUD_CDC
+    MP_USBD_BUILTIN_MSC,
+    #endif
+    #if CFG_TUD_CDC && CFG_TUD_MSC
+    MP_USBD_BUILTIN_CDC_MSC,
+    #endif
+    MP_USBD_BUILTIN_MAX,
+} mp_usbd_builtin_driver_t;
+
+#define MP_USBD_BUILTIN_DEFAULT (MP_USBD_BUILTIN_MAX - 1)
+
+// Maximum number of pending exceptions per single TinyUSB task execution
+#define MAX_PEND_EXCS 2 // TODO: namespace this
+
+typedef struct {
+    mp_obj_base_t base;
+
+    mp_obj_t desc_dev; // Device descriptor bytes
+    mp_obj_t desc_cfg; // Configuration descriptor bytes
+    mp_obj_t desc_strs; // List/dict/similar to look up string descriptors by index
+
+    // Runtime device driver callback functions
+    mp_obj_t open_itf_cb;
+    mp_obj_t reset_cb;
+    mp_obj_t control_xfer_cb;
+    mp_obj_t xfer_cb;
+
+    mp_usbd_builtin_driver_t builtin_driver;
```

Comment:
Could this just be `mp_obj_t`, and it just stores the pointer to the current built-in driver object?

Then you don't need an enum, and the logic to get/set the attribute for this is very simple.

Reply:
I think it's simpler to have the extern consts than the enum and the enum<->const mapping.

Reply:
Done, PTAL!